### ISSUE TITLE
Using Traits inside Schema's

### ISSUE THREAD

**Issue by laszlof** [zircote/swagger-php#700]
I ran into a bit of an issue, and I'm not sure if its just failure to understand proper usage, or a bug in swagger-php. I'm attempting to generate a schema based on model class that includes a couple of traits (with properties defined). 

I've tried 2 different methods to get this to work. The first, was the following:

Model Class Example:
```php
/**
 * Foo Class
 * 
 * @OA\Schema(schema="address_summary")
 */
class Foo {
  use \Foo\HasIdentityProperties; 
  use \Foo\HasAddressProperties;
}
```

HasIdentityProperties:
```php
/**
 * Object Identity Payload
 */
trait HasIdentityProperties {

  /**
   * The ID of this entity
   *
   * @var int
   * @example 12345
   * @OA\Property
   */
  public $id;

  /**
   * The identity of this entity
   *
   * @var string
   * @example Some Description Of This Entity
   * @OA\Property
   */
  public $identity;
}
```

HasAddressProperties has a similar definition as above. When attempting to use this and run it through the openapi command, I get the following:

```
$ ./vendor/bin/openapi --format yaml src/Api/
Unexpected @OA\Property(), expected to be inside @OA\AdditionalProperties, @OA\Schema, @OA\JsonContent, @OA\XmlContent, @OA\Property, @OA\Items in src/Api/Payload/HasIdentityProperties.php on line 50
Unexpected @OA\Property(), expected to be inside @OA\AdditionalProperties, @OA\Schema, @OA\JsonContent, @OA\XmlContent, @OA\Property, @OA\Items in src/Api/Portal/V1/Address/Payload/HasAddressProperties.php on line 42
Unexpected @OA\Property(), expected to be inside @OA\AdditionalProperties, @OA\Schema, @OA\JsonContent, @OA\XmlContent, @OA\Property, @OA\Items in src/Api/Portal/V1/Address/Payload/HasAddressProperties.php on line 51
Unexpected @OA\Property(), expected to be inside @OA\AdditionalProperties, @OA\Schema, @OA\JsonContent, @OA\XmlContent, @OA\Property, @OA\Items in src/Api/Portal/V1/Address/Payload/HasAddressProperties.php on line 60
Unexpected @OA\Property(), expected to be inside @OA\AdditionalProperties, @OA\Schema, @OA\JsonContent, @OA\XmlContent, @OA\Property, @OA\Items in src/Api/Portal/V1/Address/Payload/HasAddressProperties.php on line 69
Unexpected @OA\Property(), expected to be inside @OA\AdditionalProperties, @OA\Schema, @OA\JsonContent, @OA\XmlContent, @OA\Property, @OA\Items in src/Api/Portal/V1/Address/Payload/HasAddressProperties.php on line 78
Unexpected @OA\Property(), expected to be inside @OA\AdditionalProperties, @OA\Schema, @OA\JsonContent, @OA\XmlContent, @OA\Property, @OA\Items in src/Api/Portal/V1/Address/Payload/HasAddressProperties.php on line 87
Unexpected @OA\Property(), expected to be inside @OA\AdditionalProperties, @OA\Schema, @OA\JsonContent, @OA\XmlContent, @OA\Property, @OA\Items in src/Api/Portal/V1/Address/Payload/HasAddressProperties.php on line 96
Required @OA\PathItem() not found

openapi: 3.0.0
info:
  title: 'Foo API'
  version: 0.0.0
components:
  schemas:
    address_summary:
      title: 'Address Summary'
      properties:
        id:
          description: 'The ID of this entity'
          type: integer
          example: '12345'
      type: object
```

The second method I tried was just like above, except I defined the traits as their own independent schema's and used `allOf` in the summary schema. The errors went away, but this was the resulting yaml:

```
$ ./vendor/bin/openapi --format yaml src/Api/
Required @OA\PathItem() not found

openapi: 3.0.0
info:
  title: 'Foo API'
  version: 0.0.0
components:
  schemas:
    has_identity_properties:
      properties:
        identity:
          description: 'The identity of this entity'
          type: string
          example: 'Some Description Of This Entity'
      type: object
    has_address_properties:
      properties:
        address1:
          description: 'Address Line 1'
          type: string
          example: '123 Main St.'
        address2:
          description: 'Address Line 2'
          type: string
          example: 'Suite 1B.'
        city:
          description: 'City or Locality'
          type: string
          example: Southfield
        state:
          description: 'State or Province Abbreviation'
          type: string
          example: MI
        zip:
          description: 'Zip/Postal Code'
          type: string
          example: '48075'
        country:
          description: '2 Letter Country Abbreviation'
          type: string
          example: US
        type:
          description: 'Type of address'
          type: string
          default: default
          enum:
            - default
            - affiliate
            - billing
            - technical
            - rwhois
            - abuse
            - other
          example: default
      type: object
    address_summary:
      title: 'Address Summary'
      properties:
        id:
          description: 'The ID of this entity'
          type: integer
          example: '12345'
      type: object
      allOf:
        -
          $ref: '#/components/schemas/has_identity_properties'
        -
          $ref: '#/components/schemas/has_address_properties'
```

If you look at the has_identity_properties schema, you'll see its missing the "id" property. This "id" property is somehow in the "address_summary" schema. The "allOf" also looks fairly suspect.

Any thoughts on what I'm doing wrong here, or is this a bug? Thanks!

**Comment by ramsey** [comment]
I'm curious about the distinction between `$id` and `$identity`. This isn't related to your issue, but I'm curious about how these two properties differ.

**Comment by laszlof** [comment]
@ramsey In our application, `$id` is an auto-incremented int in the database (typically). Where as `$identity` is a string definition of same object. Its usually a composite identifier depending on the type of object its representing. A quick example would be that of a user, where it might contain the first and last name. Its not directly in the database, but generated from a Model method.

**Comment by ramsey** [comment]
As far as the second YAML generated, at a glance, it looks correct, except it's odd that `id` shows up under `address_summary` instead of `has_address_properties`.

**Comment by laszlof** [comment]
@ramsey Right. It looks like its taking the first property inside the first `use'd` trait and assigning it to the address summary. 

I will note that the first method would be my preference, as I'd rather not have to define the traits being used twice (both `use` and via annotation). It also is going to cause issues when we have the need to both `use` a trait, and define additional properties directly on the class itself.

**Comment by laszlof** [comment]
If its helpful, I can make a small project that illustrates the issue more clearly. It should be fairly easy to reproduce outside of our application.

**Comment by ramsey** [comment]
>  It looks like its taking the first property inside the first use'd trait and assigning it to the address summary.

If that's the case, then this is most likely an issue with the `swagger-php` library and not a lack of understanding around OpenAPI. 😁

**Comment by laszlof** [comment]
@ramsey That was my thought as well. Thanks for taking a glance :)

**Comment by laszlof** [comment]
Looks like a bug. I found [this PR](https://github.com/zircote/swagger-php/pull/576) which seems to fix the issue I encountered.

**Comment by mroeling** [comment]
I second that. Had similar issue, directly using the fork repo did fix it on my project as well. Any e.t.a. on when this will be pulled and released?

**Comment by DerManoMann** [comment]
Here is a new [PR](https://github.com/zircote/swagger-php/pull/774) to improve handling of traits. Feel free to try that and provide some feedback on the PR.

**Comment by DerManoMann** [comment]
@laszlof #782 does introduce a more consistent way of handling traits (and interfaces too). Maybe you can give that a go and provide some feedback.

**Comment by DerManoMann** [comment]
Fixed via #818, #804, #798, #782, #779
If you still see problems please raise a new issue. Thanks.
